### DISCUSSION TITLE
Solution to Pylint warning 'R0801: Similar lines in 2 files'

### DISCUSSION THREAD

**Discussion by BECATRUE** [snu-quiqcl/qiwis#41] [Ideas]
I want to discuss about Pylint warning `R0801: Similar lines in 2 files` that @Ahrrri and I are experiencing in #40 and #38.

This problem occurs because we re-use perfectly equal code such as `updateDB()`.
So I suggests a solution!

First, I think it isn't a good idea to modify `swift.app.BaseApp` because it is just a general and abstract `app`.
Then how about creating one more abstract app `ReceiveDBApp` (example) which is a subclass of `swift.app.BaseApp` and a superclass of all `app`s which implement `updateDB()`.

Like the above case, if there is another duplicated code, make a new abstract `app` class about it.

**Comment by BECATRUE** [comment]
Please feel free to leave your idea😊

**Comment by BECATRUE** [comment]
The simple structure is the below.

```python
class BaseApp:
  pass

class ReceiveDBApp(BaseApp):
  def updateDB(...):
    pass

class AnotherApp(BaseApp):
  def anotherMethod(...):
    pass

class RealApp(ReceiveDBApp, AnotherApp):
  # able to call 'updateDB()' and 'anotherMethod()'
```

**Reply by kangz12345** [reply]
In this case, how do we implement the inside of `updateDB()`? The current implementation requires, e.g., the combobox widget instances, as far as I know.

**Reply by BECATRUE** [reply]
Umm.. I didn't consider about it..

**Comment by kangz12345** [comment]
Oh, I was very uncomfortable of that, but I didn't know Pylint rejects such codes 😄

I have recognized that there are two things we can consider of:
1. The database selector combobox and its slots (e.g. `updateDB()`), which are reused in multiple apps.
2. The similar form of protocol (dict -> json -> dict; key checking, etc.).

However, I think the 1 and 2 are related quite closely.
If we provide a helper functions for such protocol boilerplate codes, the duplicate code lines could be reduced significantly (e.g., `json.loads()` error checking, key membership checking, etc.).

I will think more about where and how it should be implemented.

Or, we can compromise on this issue by increasing the `--min-similarity-lines` configuration of 
[similarities checker](https://pylint.readthedocs.io/en/latest/user_guide/configuration/all-options.html#similarities-checker).

**Reply by BECATRUE** [reply]
Wow, modifying 'similarities checker' looks very attractive!
Honestly, I think that it is too inefficient to change code structures just for meeting Pylint conditions.

**Reply by kangz12345** [reply]
I agree, then how about

1. Create an issue from this discussion (modifying pylint.yml)
2. Create a branch from the issue
3. Create a pull request
4. Merge it
5. Rebase the branches which suffer from the problem to `main`.

**Reply by BECATRUE** [reply]
It looks good! I will do so.
